### PR TITLE
Changing forward signature to explicit named variables

### PR DESCRIPTION
**Pull request by agoscinski**
As discussed in #8 the change to the forward signature

### CODE REVIEW

**Comment by agoscinski** [comment]
First draft for changing forward signature to accept neighborlists. I removed the mask in the neighborlist creation, I am the opinion that the same effect of the mask can be achieved by turning off self_interaction. At the moment the test `test_finite_differences` fails, because I have not adapted it yet.

TODOs:
- I prefer to do change the pair indices adaptively during the collate_fn evoked by the DataLoader that using a structure_offset that makes a bit the usage for single structures cumbersome. I want to change it to this style
- include tests that have more than one frame
- include tests that have nonzero cell_shifts
- change finite difference to autograd.gradcheck

**Comment by agoscinski** [comment]
I tried to put the computation of the direction vectors into tho computation of the Dataset, so it does not need to be recomputed each epoch, but there is a problem doing this when also having the option to batch arbitrary. The batched positions loose the connection to the direction vectors

```
positions --DataLoader--> batched_positions # because concatenating or slicing creates a new node in the autograd graph
   |
   |Dataset
   |
   --> direction_vectors --DataLoader--> batched_direction_vectors
```
So when backpropagating the direction vectors we cannot get the gradients from the batched positions but need the original positions. So The DataLoader returns in addition a list of positions, so one can access the gradients. @abmazitov raised the concern that autograd.grad might be compute multiple times the gradients when using a list. I registered a hook on the direction_vectors to see if they are recomputed using
```python
input_kwargs["direction_vectors"].retain_grad()
input_kwargs["direction_vectors"].register_hook(lambda _: print("USED"))
```
and could not find any recomputations of the gradients.

**Review by abmazitov — CHANGES_REQUESTED** [review]
I would suggest to rename `InMemoryNeighborList` to `InMemoryDataset` to avoid ambiguity with `TransformerNeighborList`. Everything else seems reasonable to me.

File: tests/test_spherical_expansions.py [hunk 1]
```diff
@@ -10,27 +10,35 @@
...
     device = "cpu"
     frames = ase.io.read('datasets/rmd17/ethanol1.extxyz', ':1')
-    all_species = np.unique(np.hstack([frame.numbers for frame in frames]))
-    structures = ase_atoms_to_tensordict(frames)
+    all_species = np.unique([frame.numbers for frame in frames])
     with open("tests/data/expansion_coeffs-ethanol1_0-hypers.json", "r") as f:
         hypers = json.load(f)
 
+    transformers = [TransformerNeighborList(cutoff=hypers["cutoff radius"])]
+    dataset = InMemoryNeighborList(frames, transformers)
```
[the remark this answers is not in the document]
Comment (agoscinski) [reply tests/test_spherical_expansions.py:28]: right, good catch, changed it
